hatchery/src/snapshot.rs

```diff
@@ -0,0 +1,138 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+use crate::error::Error;
+use crate::storage_helpers::{
+    combine_module_snapshot_names, snapshot_id_to_name,
+};
+use crate::Error::PersistenceError;
+use std::io::Read;
+use std::path::{Path, PathBuf};
+
+use rkyv::{Archive, Deserialize, Serialize};
+pub const SNAPSHOT_ID_BYTES: usize = 32;
+#[derive(
+    Debug,
+    PartialEq,
+    Eq,
+    Archive,
+    Serialize,
+    Deserialize,
+    PartialOrd,
+    Ord,
+    Hash,
+    Clone,
+    Copy,
+)]
+#[repr(C)]
+pub struct SnapshotId([u8; SNAPSHOT_ID_BYTES]);
+impl SnapshotId {
+    pub const fn uninitialized() -> Self {
+        SnapshotId([0u8; SNAPSHOT_ID_BYTES])
+    }
+    pub fn as_bytes(&self) -> &[u8] {
+        &self.0
+    }
+}
+impl From<[u8; 32]> for SnapshotId {
+    fn from(array: [u8; 32]) -> Self {
+        SnapshotId(array)
+    }
+}
+
+pub trait SnapshotLike {
+    fn path(&self) -> &PathBuf;
+    /// Read's snapshot's content into buffer
+    fn read(&self) -> Result<Vec<u8>, Error> {
+        let mut f = std::fs::File::open(self.path().as_path())
+            .map_err(PersistenceError)?;
+        let metadata = std::fs::metadata(self.path().as_path())
+            .map_err(PersistenceError)?;
+        let mut buffer = vec![0; metadata.len() as usize];
+        f.read(buffer.as_mut_slice()).map_err(PersistenceError)?;
+        Ok(buffer)
+    }
+}
+
+pub struct MemoryPath {
+    path: PathBuf,
+}
+
+impl MemoryPath {
+    pub fn new(path: impl AsRef<Path>) -> Self {
+        MemoryPath {
+            path: path.as_ref().to_path_buf(),
+        }
+    }
+}
+
+impl SnapshotLike for MemoryPath {
+    fn path(&self) -> &PathBuf {
+        &self.path
+    }
+}
+
+pub struct Snapshot {
+    path: PathBuf,
+    id: SnapshotId,
+}
+
+impl Snapshot {
+    pub fn new(memory_path: &MemoryPath) -> Result<Self, Error> {
+        let snapshot_id: SnapshotId = SnapshotId::from(
+            *blake3::hash(memory_path.read()?.as_slice()).as_bytes(),
```

Comment:
Shouldn't the snapshot id be somehow computed from the state itself and not just the path?

Reply:
Yes, ideally the snapshot state should be a combination of all instance ids, which themselves should be the hash of their memory excluding the argument buffer

Reply:
snapshot is computed from a file's contents the path is pointing, see the little "read()?" there? ;)

Reply:
Missed the read indeed. My bad